raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -306,10 +314,22 @@ private VoteResponseData buildVoteResponse(Errors error, boolean voteGranted) {
                 .setVoteGranted(voteGranted);
     }
 
-    private VoteResponseData handleVoteRequest(VoteRequestData request) throws IOException {
-        Optional<Exception> errorOpt = validateLeaderOnlyRequest(request.candidateId(), request.candidateEpoch());
+    /**
+     * Handle a Vote request. This API may return the following errors:
+     *
+     * - {@link Errors#BROKER_NOT_AVAILABLE} if this node is currently shutting down
+     * - {@link Errors#FENCED_LEADER_EPOCH} if the epoch is smaller than this node's epoch
+     * - {@link Errors#INCONSISTENT_VOTER_SET} if the request suggests inconsistent voter membership (e.g.
+     *      if this node or the sender is not one of the current known voters)
+     * - {@link Errors#INVALID_REQUEST} if the last epoch or offset are invalid
+     */
+    private VoteResponseData handleVoteRequest(
+        RaftRequest.Inbound requestMetadata
+    ) throws IOException {
+        VoteRequestData request = (VoteRequestData) requestMetadata.data;
+        Optional<Errors> errorOpt = validateVoterOnlyRequest(request.candidateId(), request.candidateEpoch());
```

Comment:
Related to my other comment: I feel it is a bit awkward that `validateVoterOnlyRequest` may change this voter from a leader state to an unattached follower, and then in line 335 below the check `quorum.isLeader()` would be void --- readers may expect that the `validateXXX` function would not change the state of this voter.

I'd suggest we do not call `becomeUnattachedFollower` inside the validate function, and then let the this function, the caller to decide based on its current state whether the returned error code can be handled accordingly or should be returned in the error response.

Reply:
This response is also related to my other response 🙂 . The invariant that `validateVoterOnlyRequest` and `validateLeaderOnlyRequest` provide is that if no error is returned, then the request is guaranteed to match the current epoch. I agree the state change is a little surprising given the name of the function, but maybe that just means we need a better name. The alternative is to give up the invariant and duplicate the checks. I will think about it a little more, but I'm not too eager to give this up because I felt it really did reasoning about the states simpler.